### PR TITLE
Add com.exok.Celeste64

### PR DESCRIPTION
**Pull request by brunofin**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. [**Link:**](https://github.com/ExOK/Celeste64/issues/35)
- [x] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

### CODE REVIEW

File: com.exok.Celeste64.desktop [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[Desktop Entry]
+Name=Celeste 64
+Exec=/app/bin/celeste64/bin/Release/net8.0/linux-x64/Celeste64
```
Comment (hfiguiere) [inline com.exok.Celeste64.desktop:3]: exec should be the name of the executable (not full path) that in "PATH" (/app/bin). You can make a symlink there for that purpose

File: com.exok.Celeste64.desktop [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+Type=Application
+Icon=icon.png
```
Comment (hfiguiere) [inline com.exok.Celeste64.desktop:5]: this is the wrong icon name. And it NEVER has the extension.

File: com.exok.Celeste64.metainfo.xml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  <project_license>MIT</project_license>
+  <name>Celeste 64</name>
+  <summary>Celeste 64: Fragments of the Mountain</summary>
+
+  <url type="homepage">https://exok.com/</url>
+  <url type="help">https://github.com/ExOK/Celeste64</url>
+  <url type="bugtracker">https://github.com/ExOK/Celeste64/issues</url>
+
+  <description>
+    <p>Relive the magic of Celeste Mountain alongside Madeline in this small, heartfelt 3D platformer. Created in a week(ish) by the <a href="https://exok.com/" rel="nofollow noopener" referrerpolicy="origin">Celeste team</a> to celebrate the game’s sixth anniversary 🍓✨</p>
```
Comment (hfiguiere) [inline com.exok.Celeste64.metainfo.xml:16]: `<a>` href are NOT allowed in the description here. It's not an HTML document.

File: com.exok.Celeste64.metainfo.xml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  <update_contact>bruno.k.finger_at_gmail.com</update_contact>
+
+  <!-- Contents are from https://maddymakesgamesinc.itch.io/celeste64 -->
+  <releases>
+    <release version="1.0.1" date="2024-01-30">
+      <description>
+        <ul>
+          <li>Fixed the Speedrun Timer rounding hours up (ex. so 1.6 hours would display as 2 hours)</li>
+          <li>Fixed an OpenGL error on Linux (and probably Windows)</li>
+          <li>Fixed the FMOD Symlink's on Linux (<a href="https://github.com/ExOK/Celeste64/issues/2#issuecomment-1916029872" target="_blank" rel="nofollow noopener" referrerpolicy="origin">better more permanent fix incoming</a>)</li>
```
Comment (hfiguiere) [inline com.exok.Celeste64.metainfo.xml:55]: same as above. `<a>` is not allowed here.

File: com.exok.Celeste64.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+runtime-version: '23.08'
+sdk: org.freedesktop.Sdk
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.dotnet8
+build-options:
+  append-path: "/usr/lib/sdk/dotnet8/bin"
+  append-ld-library-path: "/usr/lib/sdk/dotnet8/lib"
+  env:
+    PKG_CONFIG_PATH: "/app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet8/lib/pkgconfig"
+command: /app/bin/celeste64/bin/Release/net8.0/linux-x64/Celeste64
```
Comment (hfiguiere) [inline com.exok.Celeste64.yml:12]: comman should just be the name of the comand that in the the PATH. NOT an full path.

Comment (hfiguiere) [reply com.exok.Celeste64.yml:12]: also how would that work on aarch64?

File: com.exok.Celeste64.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  - name: celeste64
+    buildsystem: simple
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --source ./nuget-sources Celeste64.csproj"
+      - "cp -r --remove-destination /run/build/celeste64/ /app/bin/"
+      
+      - install -Dm644 ${FLATPAK_ID}.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
+      - install -Dm644 ${FLATPAK_ID}.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm644 icon.png ${FLATPAK_DEST}/share/icons/hicolor/scalable/apps/${FLATPAK_ID}.png
```
Comment (hfiguiere) [inline com.exok.Celeste64.yml:31]: a PNG is NOT scalable.

File: com.exok.Celeste64.desktop (hunk 1 above)
Comment (brunofin) [reply com.exok.Celeste64.desktop:3]: do you mean here it should be only `Celeste64`? I could change the build steps to get the contents of `linux-x64` in the root of `/app/bin` in this case probably. Would that work?

File: com.exok.Celeste64.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  - --socket=wayland
+  - --socket=fallback-x11
+
+modules:
+  - name: celeste64
+    buildsystem: simple
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --source ./nuget-sources Celeste64.csproj"
+      - "cp -r --remove-destination /run/build/celeste64/ /app/bin/"
```
[the remark this answers is not in the document]
Comment (brunofin) [reply com.exok.Celeste64.yml:27]: yeah I guess that's what causes the issue with the comment above with `Exec=`?

File: com.exok.Celeste64.yml (hunk 6 above)
Comment (brunofin) [reply com.exok.Celeste64.yml:31]: Could you help me with this one? What is the correct path for a PNG icon?

File: com.exok.Celeste64.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (hfiguiere) [reply com.exok.Celeste64.yml:27]: and why in `/app/bin` ?

**Comment by brunofin** [comment]
@hfiguiere Thank you very much for your review. I've updated the MR to address your comments.

File: com.exok.Celeste64.yml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      - "dotnet publish -c Release --runtime $RUNTIME --source ./nuget-sources Celeste64.csproj"
+      - "cp -r --remove-destination /run/build/celeste64/bin/Release/net8.0/$RUNTIME/publish/ /app/${FLATPAK_ID}"
+      - "cp -r --remove-destination /run/build/celeste64/Content/ /app/${FLATPAK_ID}/Content"
+      - "ln -s /app/${FLATPAK_ID}/Celeste64 /app/bin/celeste64"
+      
+      - install -Dm644 ${FLATPAK_ID}.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
+      - install -Dm644 ${FLATPAK_ID}.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm644 ${FLATPAK_ID}.png ${FLATPAK_DEST}/share/icons/hicolor/128x128/apps/${FLATPAK_ID}.png
+      
+      - desktop-file-edit --set-icon="${FLATPAK_ID}" ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
```
Comment (hfiguiere) [inline com.exok.Celeste64.yml:41]:
```suggestion
```

File: com.exok.Celeste64.yml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      - desktop-file-edit --set-key="X-Flatpak-RenamedFrom" --set-value="com.exok.Celeste64.desktop" ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
```
Comment (hfiguiere) [inline com.exok.Celeste64.yml:42]:
```suggestion
```

Not sure why you think you need that.

File: com.exok.Celeste64.yml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    build-commands:
+      - "install-sdk.sh"
+      - "dotnet publish -c Release --runtime $RUNTIME --source ./nuget-sources Celeste64.csproj"
+      - "cp -r --remove-destination /run/build/celeste64/bin/Release/net8.0/$RUNTIME/publish/ /app/${FLATPAK_ID}"
+      - "cp -r --remove-destination /run/build/celeste64/Content/ /app/${FLATPAK_ID}/Content"
+      - "ln -s /app/${FLATPAK_ID}/Celeste64 /app/bin/celeste64"
+      
+      - install -Dm644 ${FLATPAK_ID}.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
+      - install -Dm644 ${FLATPAK_ID}.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm644 ${FLATPAK_ID}.png ${FLATPAK_DEST}/share/icons/hicolor/128x128/apps/${FLATPAK_ID}.png
```
Comment (hfiguiere) [inline com.exok.Celeste64.yml:39]: the image is 150x150. (which is not a valid size either)

File: com.exok.Celeste64.yml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  append-path: "/usr/lib/sdk/dotnet8/bin"
+  append-ld-library-path: "/usr/lib/sdk/dotnet8/lib"
+command: celeste64
+finish-args:
+  - --device=all
+  - --env=XCURSOR_PATH=/run/host/user-share/icons:/run/host/share/icons
+  - --share=ipc
+  - --socket=pulseaudio
+  - --socket=wayland
+  - --socket=fallback-x11
```
Comment (hfiguiere) [inline com.exok.Celeste64.yml:25]: does it really support wayland?

**Comment by hfiguiere** [comment]
bot, build com.exok.Celeste64

**Comment by brunofin** [comment]
Seems the aarm64 build failed. I can try to include the missing nuget packages for this architecture too but honestly I am not even sure if the app supports arm64.

Is there a way to flag in the manifest supported architectures?

**Comment by xordspar0** [comment]
> I am not even sure if the app supports arm64.

Someone did get it working on arm64: https://github.com/ExOK/Celeste64/issues/23

**Comment by brunofin** [comment]
@hfiguiere thank you once again for your time reviewing. I believe I have addressed all your comments again.

File: com.exok.Celeste64.metainfo.xml [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+      <image type="source">https://raw.githubusercontent.com/brunofin/flathub/celeste64/screenshots/3.png</image>
+    </screenshot>
+  </screenshots>
+
+  <categories>
+    <category>Game</category>
+  </categories>
+  
+  <content_rating type="oars-1.0">
+  </content_rating>
```
Comment (bbhtt) [inline com.exok.Celeste64.metainfo.xml:39]:
```suggestion
  <content_rating type="oars-1.1" />
```
https://docs.flathub.org/docs/for-app-authors/appdata-guidelines/#oars-information

File: com.exok.Celeste64.yml [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+build-options:
+  arch:
+    x86_64:
+      env:
+        RUNTIME: "linux-x64"
+# support for aarm64 will be added in the future
+#    aarch64:
+#      env:
+#        RUNTIME: "linux-arm64"
+  prepend-pkg-config-path: "/app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet8/lib/pkgconfig"
```
Comment (bbhtt) [inline com.exok.Celeste64.yml:16]:
```suggestion
  prepend-pkg-config-path: "/usr/lib/sdk/dotnet8/lib/pkgconfig"
```

prepend means you don't need to add the whole of existing paths

File: com.exok.Celeste64.yml [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  - --env=XCURSOR_PATH=/run/host/user-share/icons:/run/host/share/icons
+  - --share=ipc
+  - --socket=pulseaudio
+  - --socket=x11
+
+modules:
+  - name: celeste64
+    buildsystem: simple
+    only-arches:
+      - "x86_64"
```
Comment (bbhtt) [inline com.exok.Celeste64.yml:31]: You need this too https://docs.flathub.org/docs/for-app-authors/maintenance#limiting-the-set-of-architectures-to-build-on

**Review by bbhtt — COMMENTED** [review]
Can desktop file, icon, metainfo file and screenshots be put upstream?

**Comment by brunofin** [comment]
> Can desktop file, icon, metainfo file and screenshots be put upstream?

Hi, unfortunately I think they can't. I am not the owner or a contributor at mainstream, I only introduced them the idea of publishing Celeste 64 to Flathub which they thought it was awesome but didn't have the bandwidth to do it themselves, so they happily agreed with allowing me to do that.

https://github.com/ExOK/Celeste64/issues/35

Having that said, moving them upstream would cause making changes to flatpak and Flathub specific files description screenshots and etc hard for me to maintain it without write access to upstream, so I would appreciate if they could stay here.

If that's still a blocker, I could create a new repo owned by me to host those files and import them as a git source in the flatpak manifest if that's a possibility.

**Comment by brunofin** [comment]
@hfiguiere and @bbhtt  thank you once again for your time reviewing. I believe I have addressed all your comments again.

**Comment by hfiguiere** [comment]
> > Can desktop file, icon, metainfo file and screenshots be put upstream?
> 
> Hi, unfortunately I think they can't. I am not the owner or a contributor at upstream,

You can submit pull request, can't you?
The thing is not shipping .desktop and icons for such an application is a bug.
None of these are specific flathub. The appstream file is also used independently of Flathub to provide information on the application, discoverable.


> Having that said, moving them upstream would cause making changes to flatpak and Flathub specific files description screenshots and etc hard for me to maintain it without write access to upstream, so I would appreciate if they could stay here.

As I said they are not specific to flatpak.

**Comment by brunofin** [comment]
I'

> > > Can desktop file, icon, metainfo file and screenshots be put upstream?
> > 
> > 
> > Hi, unfortunately I think they can't. I am not the owner or a contributor at upstream,
> 
> You can submit pull request, can't you? The thing is not shipping .desktop and icons for such an application is a bug. None of these are specific flathub. The appstream file is also used independently of Flathub to provide information on the application, discoverable.
> 
> > Having that said, moving them upstream would cause making changes to flatpak and Flathub specific files description screenshots and etc hard for me to maintain it without write access to upstream, so I would appreciate if they could stay here.
> 
> As I said they are not specific to flatpak.

I've moved the files upstream. For now that means I created a fork of the original repository and aded the files there. I've submitted a PR explaining why those files are needed but I am not sure when or if they will be merged. If they do, I can update update th manifest once again pointing to the original repository. I hope this works for the time being?

**Comment by brunofin** [comment]
The owner of upstream said they preferred the files to not be there. They said they think it makes sense the files stays in my fork, so it's a matter of preference here, whether the git source points to my fork, or to the original.repos but the files stay here.

In case you prefer the 1st option (my fork) everything is ready, otherwise please let me know if you prefer git source pointing to the original repo but desktop files and etc here.

Thanks.

File: com.exok.Celeste64.yml [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+      - "cp -r --remove-destination /run/build/celeste64/Content/ /app/${FLATPAK_ID}/Content"
+      - "ln -s /app/${FLATPAK_ID}/Celeste64 /app/bin/celeste64"
+      
+      - install -Dm644 .flatpak/${FLATPAK_ID}.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
+      - install -Dm644 .flatpak/${FLATPAK_ID}.desktop ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm644 .flatpak/${FLATPAK_ID}.png ${FLATPAK_DEST}/share/icons/hicolor/128x128/apps/${FLATPAK_ID}.png
+
+    sources:
+      - type: "git"
+        url: "https://github.com/brunofin/Celeste64.git"
```
Comment (bbhtt) [inline com.exok.Celeste64.yml:46]: Please don't switch it to a fork. You can add direct links to desktop file icon as secondary sources.

**Comment by bbhtt** [comment]
> In my honest opinion I believe those files do not belong in upstream but only here, especially given I see a large amount of accepted submissions where those files are present there too instead of upstream.

https://docs.flathub.org/docs/for-app-authors/requirements#acceptable-but-should-be-submitted-upstream

It's accepted because a) proprietary b) unresponsive upstream c) uncooperative upstream etc.

Desktop file, appdata can have translations and regularly needs to be updated with release notes. The best place to contribute them and translations is upstream. Also they are supposed to be used by other distribution methods like snap, regular linux packages etc. This makes everything fragmented.

In any case, I think the flathub repo gives them more visibility than your fork.

**Comment by brunofin** [comment]
@bbhtt I've switched back the code source to the original repo.

As for the other files, I've created a secondary repo dedicated for them.

**Comment by brunofin** [comment]
@bbhtt sorry I realize now I probably misunderstood your comment here:

> Please don't switch it to a fork. You can add direct links to desktop file, icon etc. as secondary sources.

I understood you wanted me to create a secondary git source, but then I read now:

> In any case, I think the flathub repo gives them more visibility than your fork.

Saying you were ok with having those files back in here.  I fixed that.

The PR should be ready now. Please let me know if further request.

**Comment by bbhtt** [comment]
You can add direct links to the files after committing and pushing them in this PR. Anyways, it's your choice, I said I think the Flathub repo gives it more visibility.

**Comment by bbhtt** [comment]
bot, build com.exok.Celeste64

**Comment by barthalion** [comment]
bot, build com.exok.Celeste64

**Pull request closed without merging**